### PR TITLE
Commerce 5797 dxp 3of5 final

### PR DESCRIPTION
**Pull request by garofalog**
Hello you all, here it's 3 of 5 story about Shop By Diagram.
We documented the features [here](https://issues.liferay.com/browse/COMMERCE-5797).
nb. Features on AdminTooltip part (popping out, editing pins, deleting pins) was removed at this time and Import-File-Table about mapped product will be added in future.

The main features added are:
1. isAdmin prop to choose if we're using this in Admin or Shopping Experience side
2. Was added dblclick-add-pin feature to make the admin user to add a new pin

### CODE REVIEW

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:relevant

**Comment by igorbeslic** [comment]
ci:forward

**Comment by marco-leo** [comment]
ci:reopen

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:relevant

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramFooter.js [hunk 1]
```diff
@@ -12,26 +12,43 @@
...
 		<div className="d-flex diagram-footer justify-content-end mt-3">
+			{isAdmin && (
+				<ClayButton
+					className="mr-3"
+					onClick={() => setAddPinHandler(true)}
+				>
+					<span className="inline-item inline-item-before">
+						<ClayIcon symbol="pin" />
+					</span>
+					{Liferay.Language.get('add-pin')}
```
Comment (kresimir-coko) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramFooter.js:49]: Empty line between siblings

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js [hunk 2]
```diff
@@ -15,124 +15,148 @@ import ClayDropDown from '@clayui/drop-down';
...
-								step={1}
-								title={Liferay.Language.get('custom-colors')}
-								useNative={newPinSettings.colorPicker.useNative}
-								value={addNewPinState.fill}
-							/>
-						</ClayDropDown.Group>
-					</ClayDropDown.ItemList>
+			{isAdmin && (
+				<div className="d-flex text-align-center">
+					<label className="align-middle my-auto">
```
Comment (kresimir-coko) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js:54]: What is `my-auto`?

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js [hunk 3]
```diff
@@ -15,124 +15,148 @@ import ClayDropDown from '@clayui/drop-down';
...
+												fill: addNewPinState.fill,
+												radius: item,
+											})
+										}
+										value={addNewPinState.radius}
+									/>
+
+									<ClayButton
+										block
+										displayType="primary"
```
Comment (kresimir-coko) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js:149]: `primary`  is the default `displayType`, so no need to define it here

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js [hunk 4]
```diff
@@ -15,124 +15,148 @@ import ClayDropDown from '@clayui/drop-down';
...
+									>
+										{Liferay.Language.get('add-pin')}
+									</ClayButton>
+								</ClayForm.Group>
+							</ClayForm>
+						</ClayDropDown.Caption>
+					</ClayDropDown>
+				</div>
+			)}
 			<ClayButton
```
Comment (kresimir-coko) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js:160]: Empty line

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js [hunk 5]
```diff
@@ -31,72 +33,86 @@ const PIN_ATTRIBUTES = [
...
+				(d3event.offsetX - transform.current.x) / transform.current.k;
+			const y =
+				(d3event.offsetY - transform.current.y) / transform.current.k;
+
+			setCpins(
+				cPins.concat({
+					cx: x,
+					cy: y,
+					draggable: true,
+					fill: '#' + addNewPinState.fill,
```
Comment (kresimir-coko) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js:93]: Template literal

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js [hunk 6]
```diff
@@ -33,10 +33,7 @@ const NavigationButtons = ({moveController, position}) => {
 			<div className="box left" onClick={() => moveController('left')}>
 				<ClayIcon className="icon" symbol="angle-down" />
 			</div>
-			<div
-				className="bottom box"
-				onClick={() => moveController('bottom')}
-			>
+			<div className="bottom box" onClick={() => moveController('down')}>
 				<ClayIcon className="icon" symbol="angle-down" />
 			</div>
```
Comment (kresimir-coko) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js:38]: Empty line missing between 35 and 36. Is there a better semantic element instead these `div`s here?

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/PinUtils.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ * distribution rights of the Software.
+ */
+
+export const addPin = (cPins, setCpins, addNewPinState) => {
+	setCpins(
+		cPins.concat({
+			cx: 50,
+			cy: 50,
+			draggable: true,
+			fill: '#' + addNewPinState.fill,
```
Comment (kresimir-coko) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/PinUtils.js:18]: Template literals

**Review by kresimir-coko — CHANGES_REQUESTED** [review]
Looks good, can't see any blockers but some SF-ish things to do still.

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js (hunk 2 above)
Comment (garofalog) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js:54]: align vertically

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramFooter.js (hunk 1 above)
Comment (garofalog) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramFooter.js:49]: sf didn't add it automatically

Comment (kresimir-coko) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramFooter.js:49]: Doesn't matter, still need to have it. A lot of things aren't automatic but we still need to do them

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:relevant

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js [hunk 8]
```diff
@@ -317,15 +281,14 @@ const ImagePins = ({
 		}
 
 		if (isAdmin) {
-			select('#newPin').on('click', handleAddPin);
+			select('#newPin').on('click', addPin);
 		}
 		// eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (markocikos) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js:286]: Why do we need this `eslint-disable`?

**Review by markocikos — COMMENTED** [review]
@garofalog Looks good, some minor comments inline.

This part of code already merged, but I'm curious, what's the purpose of [these kind of lines](https://github.com/liferay/liferay-portal/blob/master/modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js#L274)?
```js
.attr('id', (attr) => attr.id)
```
I'm not too familiar with `d3` API, but this looks to me like we are not doing anything. Aren't we setting `id` attribute value to itself? A sidenote, `id` in linked file is set again likely needlessly [again](https://github.com/liferay/liferay-portal/blob/master/modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js#L281).

**Comment by garofalog** [comment]
> @garofalog Looks good, some minor comments inline.
> 
> This part of code already merged, but I'm curious, what's the purpose of [these kind of lines](https://github.com/liferay/liferay-portal/blob/master/modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js#L274)?
> 
> ```js
> .attr('id', (attr) => attr.id)
> ```
> 
> I'm not too familiar with `d3` API, but this looks to me like we are not doing anything. Aren't we setting `id` attribute to value it already is? A sidenote, `id` in linked file is set again likely needlessly [again](https://github.com/liferay/liferay-portal/blob/master/modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js#L281).

We need it because it let every circle to have a unique id that will interact with a list through it

**Comment by markocikos** [comment]
> We need it because it let every circle to have a unique id that will interact with a list through it

Let me rephrase the question more clearly. I'm not asking about `id` attribute specifically, but this type of code:
```js
.attr('cx', (attr) => attr.cx)
.attr('cy', (attr) => attr.cy)
.attr('id', (attr) => attr.id)
.attr('label', (attr) => attr.label)
.attr('fill', (attr) => attr.fill)
.attr('linked_to_sku', (attr) => attr.linked_to_sku)
.attr('quantity', (attr) => attr.quantity)
.attr('r', (attr) => attr.r)
.attr('sku', (attr) => attr.sku)
.attr('id', (attr) => attr.id)
```

To me, this looks like `element.setAttribute('foo', element.getAttribute('foo'))`.

What is the `attr` object in callback?

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js [hunk 9]
```diff
@@ -31,72 +33,86 @@ const PIN_ATTRIBUTES = [
...
+
+	const {
+		addNewPinState,
+		addPinHandler,
+		cPins,
+		isAdmin,
+		removePinHandler,
+		resetZoom,
+		selectedOption,
+		setAddPinHandler,
```
Comment (diegonvs) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js:58]: Hey @garofalog, What about using [`useReducer` for creating a local store and then just get `dispatch` function](https://pt-br.reactjs.org/docs/hooks-reference.html#usereducer) in DiagramCommonContext?

There is an issue when passing state setters over React context, see:

From https://reactjs.org/docs/hooks-reference.html#usereducer:

> useReducer is usually preferable to useState when you have complex state logic that involves multiple sub-values or when the next state depends on the previous one. useReducer also lets you optimize performance for components that trigger deep updates because you can pass dispatch down instead of callbacks.

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js [hunk 10]
```diff
@@ -24,18 +24,24 @@ const NavigationButtons = ({moveController, position}) => {
 				top: position.top,
 			}}
 		>
-			<div className="box top" onClick={() => moveController('up')}>
+			<button className="box top" onClick={() => moveController('up')}>
 				<ClayIcon className="icon" symbol="angle-left" />
-			</div>
-			<div className="box right" onClick={() => moveController('right')}>
+			</button>
```
[the remark this answers is not in the document]
Comment (diegonvs) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js:29]: I think He didn't used ClayButtonWithIcon to no have `btn-*` classes, just `box top` classes. However, I think @garofalog, you can use ClayButtonWithIcon with displayType: **unstyled** and pass `box top` as className. It will look like:

```jsx
<ClayButtonWithIcon
  className="box top"
  displayType="unstyled"
  onClick={() => moveController('up')}
  symbol="angle-left"
/>
```

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/PinUtils.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+ * The contents of this file are subject to the terms of the Liferay Enterprise
+ * Subscription License ("License"). You may not use this file except in
+ * compliance with the License. You can obtain a copy of the License by
+ * contacting Liferay, Inc. See the License for the specific language governing
+ * permissions and limitations under the License, including but not limited to
+ * distribution rights of the Software.
+ */
+
+export const addPin = (cPins, setCpins, addNewPinState) => {
+	setCpins(
```
Comment (diegonvs) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/PinUtils.js:13]: I think if you move to a redux-like styles and prevent using setters as mentioned in https://github.com/liferay-frontend/liferay-portal/pull/1242/files#r669765413, it will look more easier to understand.

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js [hunk 12]
```diff
@@ -15,123 +15,143 @@ import ClayDropDown from '@clayui/drop-down';
...
-								<label htmlFor="slider">
-									{Liferay.Language.get('custom-radius')}
-								</label>
+						<ClayDropDown.ItemList>
+							<ClayDropDown.Group
+								header={Liferay.Language.get('select-radius')}
+							>
+								{RADIUS_CHOICE.map((item, i) => (
+									<ClayDropDown.Item
+										key={i}
```
Comment (diegonvs) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js:109]: Instead of using numeric indexes, is there an unique identifier for each `item`? If so, I think you can use it :)

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js [hunk 13]
```diff
@@ -15,123 +15,143 @@ import ClayDropDown from '@clayui/drop-down';
...
-								<ClayDropDown.Item
-									key={i}
-									onClick={() => {
+					<ClayDropDown
+						active={active}
+						className="my-auto"
+						onActiveChange={setActive}
+						trigger={
+							<ClayButton
+								alt={Liferay.Language.get(
```
Comment (diegonvs) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js:64]: I think it can be an `aria-label` isn't? I never saw using `alt` attrs for buttons. See https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js [hunk 14]
```diff
@@ -15,123 +15,143 @@ import ClayDropDown from '@clayui/drop-down';
...
-	newPinSettings,
-	radiusChoice,
-	setAddNewPinState,
-	setAddPinHandler,
-}) => {
+import DiagramCommonContext from './context/DiagramCommonContext';
+
+const RADIUS_CHOICE = [
+	{
+		label: 'Small',
```
Comment (diegonvs) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/DiagramHeader.js:24]: Could We use `Liferay.Language.get` here?

**Comment by garofalog** [comment]
> > We need it because it let every circle to have a unique id that will interact with a list through it
> 
> Let me rephrase the question more clearly. I'm not asking about `id` attribute specifically, but this type of code:
> 
> ```js
> .attr('cx', (attr) => attr.cx)
> .attr('cy', (attr) => attr.cy)
> .attr('id', (attr) => attr.id)
> .attr('label', (attr) => attr.label)
> .attr('fill', (attr) => attr.fill)
> .attr('linked_to_sku', (attr) => attr.linked_to_sku)
> .attr('quantity', (attr) => attr.quantity)
> .attr('r', (attr) => attr.r)
> .attr('sku', (attr) => attr.sku)
> .attr('id', (attr) => attr.id)
> ```
> 
> To me, this looks like `element.setAttribute('foo', element.getAttribute('foo'))`.
> 
> What is the `attr` object in callback?

Hello, it's seems like that but it's the way d3 add attributes to elements. We need all that datas at this time because we use them to update the state when a pin changes its position or information related to it

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/css/diagram.scss [hunk 15]
```diff
@@ -95,11 +95,13 @@
...
-		
-		:focus {
-			outline: none;
+
+		:focus,
+		:active {
+			border: none !important;
 			box-shadow: none;
-			outline-color: transparent !important;
+			outline: 0 !important;
```
[the remark this answers is not in the document]
Comment (garofalog) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/css/diagram.scss]: This was just a try

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (garofalog) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js:29]: i tried all the ways you have proposed. But replacing divs with button, ClayButton or ClayButtonWithIcon the component style look different and the margin on :focus keeps showing also in different ways and that's no even way to add an icon to the svg contained in button

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js (hunk 9 above)
Comment (garofalog) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js:58]: Hello @diegonvs, and thank you. i was using react context not trying to replace reducers with it, it was just a try but i didn't want to solve anything but only clean up the code.
You're right, using useReducer should resolve a bit this handler's mess (i started doing it simple) so i used that boolean states to call events outside ImagePins component. Also if it works fine i was trying yet to replace in a different way you said that behaviour. Using reducers seems fine to me also because we are facing different topics about others related components and this kind of refactor could help us even because we should update a liferay state outiside react components to make our three react-component comunicate in LR.

If something it's not clear please ask me, we decided this behaviour more or less, today.
Going to the end we want to refactor in this way but we can't do it in this pr, we'll create a subtask to work on it later.
So i'll remove the useContext if seems to be not useful.

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (garofalog) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js:29]: Hello guys i retried to make the things work with ClayButtons but it takes a lot of work to override the clay style and add my styles properly to the svg.
Could i solve addin a `type="button"` to divs?

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js [hunk 16]
```diff
@@ -23,25 +23,32 @@ const NavigationButtons = ({moveController, position}) => {
...
 			}}
+			type="button"
 		>
-			<button className="box top" onClick={() => moveController('up')}>
-				<ClayIcon className="icon" symbol="angle-left" />
-			</button>
-			<button
-				className="box right"
-				onClick={() => moveController('right')}
+			<div
```
Comment (diegonvs) [inline modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js:28]: Hey @garofalog, why are you using a div instead of a button? If it's something related to styles I think you can try use an `<a type="button"` there :)

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (diegonvs) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js]: Sorry, I didn't saw this comment 😅 

I suggested using an anchor element in https://github.com/liferay-frontend/liferay-portal/pull/1242#discussion_r670500090

File: modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js (hunk 16 above)
Comment (garofalog) [reply modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/NavigationButtons.js:28]: sorry. <a type="button"> works fine.
Can i keep going on this pr? i opened some subtask to work about what we said yet. [here](https://issues.liferay.com/browse/COMMERCE-6910), [here](https://issues.liferay.com/browse/COMMERCE-6907), [here](https://issues.liferay.com/browse/COMMERCE-6906) and [https://issues.liferay.com/browse/COMMERCE-6905](url) and [here](https://issues.liferay.com/browse/COMMERCE-6909)

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:relevant

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:relevant

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:sf

**Comment by garofalog** [comment]
ci:test:relevant

**Comment by garofalog** [comment]
ci:forward
